server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -352,19 +352,18 @@ private synchronized void onMergeFailure(Exception exc) {
                 circuitBreakerBytes = 0;
             }
             failure.compareAndSet(null, exc);
-            MergeTask task = runningTask.get();
-            runningTask.compareAndSet(task, null);
-            onPartialMergeFailure.accept(exc);
-            List<MergeTask> toCancels = new ArrayList<>();
+            final MergeTask task = runningTask.getAndSet(null);
```

Comment:
I think it makes sense to move clearing the running task a couple of lines down, so it's closer to the place where it gets added to the list of releasable resources.

```java
final MergeTask task = runningTask.getAndSet(null);
if (task != null) {
    toCancels.add(task::cancel);
}
```

Reply:
++